### PR TITLE
Pydantic config

### PR DESCRIPTION
**Pull request by maringuu**
Cherry picked from #749.
These commits don't change any old behavior. They just introduce new functionality.
One user facing change is that the validity of main.cfg is checked, which is especially useful after #768.

### CODE REVIEW

**Review by jstucke — COMMENTED** [review]
The validation seems to work fine with feedback that is easy to understand. Would it be possible to add a test for the validation to make sure everything works as intended?

File: src/config.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    parsed_sections['unpack']['whitelist'] = _parse_comma_separated_list(parser._sections['unpack']['whitelist'])
+    for plugin_set in parsed_sections['default-plugins']:
+        parsed_sections['default-plugins'][plugin_set] = _parse_comma_separated_list(parser._sections['default-plugins'][plugin_set])
+
+    # hyphens may not be contained in identifiers
+    # plugin names may also not contain hyphens, so this is fine
+    _replace_hyphens_with_underscores(parsed_sections)
+
+    global _cfg
+    global _configparser_cfg
```
[the remark this answers is not in the document]
Comment (maringuu) [reply src/config.py:120]: I mean sure it is an anti-pattern but for this usecase it is perfectly fine I think.
As both globals variables are prefixed with an underscore anyone using them knows that they are private and should not be used.
Also both variables will only ever be used used in the `load_config` function.

File: src/helperFunctions/program_setup.py [hunk 2]
```diff
@@ -103,10 +105,19 @@ def _load_config(args):
...
     if args.log_file is not None:
         config['logging']['logfile'] = args.log_file
+        configparser_cfg['logging']['logfile'] = args.log_file
+        cfg.logging.logfile = args.log_file
     if args.log_level is not None:
         config['logging']['loglevel'] = args.log_level
+        configparser_cfg['logging']['loglevel'] = args.log_level
+        cfg.logging.loglevel = args.log_level
+
+    assert configparser_cfg._sections == config._sections
```
[the remark this answers is not in the document]
Comment (maringuu) [reply src/helperFunctions/program_setup.py:121]: Oh, right this should rather be a test than an assert in productoin code.

**Comment by maringuu** [comment]
I added two tests.  When usage of the module reveals more test cases I'm happy to add more.
 Right now I could not think of any other testcases.

**Comment by maringuu** [comment]
Rebased and changed destination branch.

**Comment by jstucke** [comment]
Docs CI fails because "pydantic" needs to be added to `autodoc_mock_imports` in `docsrc/conf.py`

**Comment by maringuu** [comment]
Fixed, I didn't run the CI again since only one line changed.

**Comment by maringuu** [comment]
The CI fails for some mysterious reason. The tests cant be collected, due to seemingly unrelated reasons. Locally everything works.

**Review by jstucke — APPROVED** [review]

**Pull request merged**